### ISSUE TITLE
A sleeper error.

### ISSUE THREAD

**Issue by 177jiang** [Minep/lunaix-os#13]
### If we run the code like this:

```c

void __USER__ _lxinit_main() {

  pid_t pid;
  for(size_t i=0; i<7; ++i){
    if(!(pid=fork())){
      while(1){
        sleep(1);
        kprintf("%d\n", i);
      }
    }
  }
  spin();
}
```
### I did some modifications in my code:
```c
void check_sleepers(){

    struct task_struct *init_task = sched_ctx.tasks + 1;
    struct task_struct *pos, *n;
    time_t now = clock_systime();

    list_for_each(pos, n, &init_task->timer.sleepers, timer.sleepers){

        if(TASK_TERMINATED(pos->state)){
            goto __delete_slp;
        }

        time_t *wtim = &pos->timer.wakeup;
        time_t *atim = &pos->timer.alarm;

        if(*wtim && now >= *wtim){
            *wtim      = 0;
            pos->state = TASK_STOPPED;
        }

        if(*atim && now >= *atim){
            *atim = 0;
        }

        if(!*wtim && !*atim){
    __delete_slp:
            list_delete(&pos->timer.sleepers);
        }
    }

}

```  
### It could solve this problem  preliminarily. I'm not sure if it has other errors.

**Comment by Minep** [comment]
Thank you for the head up. It is indeed a serious problem if it comes true! However, I have experimented your test case in my latest edition (i,e, 313093773e81eb89ad3c44b048a79bbcd6e94099) and it turns out to be expected, there is no infinite loop as your mention. 

However (of course, there is always a twist :P), when testing on that particular commit you stated, it does reproduce the issue. It indicate that is that the issue has been resolved somewhere in the future, but I don't recall (though I am sure not similar to your workaround), I might need to review the history to give you more information about this.

**Comment by Minep** [comment]
Alright, I have checked the commit history and it turns out that my 'successful' test on latest edition is no more than a result of fortune. This issue is in fact never get a chance to be resolved. It just rest in there and get triggered in a certain condition (like your test case).

Now, let's get back to the issue itself and try to analyse. First, let me refresh your memory about the sleep syscall:

```c
__DEFINE_LXSYSCALL1(unsigned int, sleep, unsigned int, seconds)
{
    if (!seconds) {
        return 0;
    }

    if (__current->sleep.wakeup_time) {
        return (__current->sleep.wakeup_time - clock_systime()) / 1000U;
    }

    __current->sleep.wakeup_time = clock_systime() + seconds * 1000;
    llist_append(&sched_ctx._procs[0].sleep.sleepers,
                 &__current->sleep.sleepers);

    __current->intr_ctx.registers.eax = seconds;
    __current->state = PS_BLOCKED;
    schedule();
}
```

When a first sleep is executed (pid=1), kernel add a 'sleeper' to the sleeper chain lead by root process (pid=0), and pause the process (PS_RUNNING -> PS_BLOCK) schedule. After 1s elapsed, the sleeper's timer expired and hence it rest to 0 and wake the process (PS_BLOCK -> PS_STOPPED). So far so good right? But the tricky thing comes to play when the sleep(2) syscall get called again. The syscall try to append the process's sleeper into chain, but it has already been there and this unguarded appending will disrupt the chain continuity and lead to undefined behavior. That's why the `check_sleeper` get stucked as it try to iterate through a broken chain!

Ok, let's get back to your proposed solution. You indeed capture the first half of the cause by realizing that the sleeper is 'never' get deleted (in fact, it is a delayed deletion) and you try to delete it as soon as it expire. That's a reasonable approach but you do not consider the case where alarm and sleep both get set (since a sleeper's node contains both alarm and sleep's timer), for example, consider the following setup:
```c
alarm(10);
sleep(3); 
```
If using your approach, the sleeper node will get removed soon after sleep's timer expired but **you will lost your alarm!** That is the reason we delay our deletion to next round of checking (so the conditions will be respected).

Thus, the correct solution is to guard the 'append' operation in sleep syscall as follow, to ensure that we only append it to chain when it is not there, as what we already did in alarm syscall:
```c
if (llist_empty(&__current->sleep.sleepers)) {
        llist_append(&sched_ctx._procs[0].sleep.sleepers,
                 &__current->sleep.sleepers);
}
```
Which I will be post a fix as soon as possible.

Good job! It is a great catch!
